### PR TITLE
feat: Add psp-account-compression program with address space

### PR DESCRIPTION
**Pull request by vadorovsky**
Add a new psp-account-compression program which is going to contain all instructions related to generalized account compression.

As the first functionality, add address space Merkle tree - an indexed Merkle tree with corresponding program instructions, which stores address spaces (as ranges). It works the same way that we envisioned nullifier Merkle tree to work.

### CODE REVIEW

File: merkle-tree/indexed/Cargo.toml [hunk 1]
```diff
@@ -5,6 +5,7 @@ edition = "2021"
 
 [dependencies]
 ark-ff = "0.4"
+borsh = { version = "0.10" }
```
Comment (ananas-block) [inline merkle-tree/indexed/Cargo.toml:8]: why don't you use anchor deserialize?
 I have had dependency conflicts trying to use both borsh and anchor

File: merkle-tree/indexed/src/array.rs [hunk 2]
```diff
@@ -1,7 +1,16 @@
 use std::{cmp::Ordering, marker::PhantomData};
 
-use ark_ff::BigInteger;
+use ark_ff::{BigInteger, BigInteger256};
+use borsh::{BorshDeserialize, BorshSerialize};
 use light_hasher::{errors::HasherError, Hasher};
+use light_utils::be_bytes_to_bigint;
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct RawIndexingElement<const N: usize> {
```
Comment (ananas-block) [inline merkle-tree/indexed/src/array.rs:9]: Is this to enable better offchain deserialization?

File: programs/address/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        Ok(())
+    }
+
+    pub fn initialize_merkle_tree(_ctx: Context<InitializeMerkleTree>) -> Result<()> {
+        Ok(())
+    }
+
+    pub fn insert_address(
+        ctx: Context<InsertAddress>,
+        addresses: [[u8; 32]; NR_ADDRESSES],
```
Comment (ananas-block) [inline programs/address/src/lib.rs:41]: We will want to make this dynamic, probably slice of [u8;32] is best?

File: programs/address/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+pub mod errors;
+pub mod state;
+
+declare_id!("5QPEJ5zDsVou9FQS3KCauKswM3VwBEBu4dpL9xTqkWwN");
+
+#[constant]
+pub const NR_ADDRESSES: usize = 2;
+#[constant]
+pub const MERKLE_TREE_HEIGHT: usize = 22;
```
Comment (ananas-block) [inline programs/address/src/lib.rs:20]: we will want height 40

File: merkle-tree/indexed/Cargo.toml (hunk 1 above)
Comment (vadorovsky) [reply merkle-tree/indexed/Cargo.toml:8]: Very funny reason - adding anchor as a dependency changes the behavior of `Result` - you have to stick to Anchor's `#[error_code]` and can't use `thiserror`.

Dependency conflicts happen if you try to use too old or too new borsh, but using 0.10.3 works fine.

File: merkle-tree/indexed/src/array.rs (hunk 2 above)
Comment (vadorovsky) [reply merkle-tree/indexed/src/array.rs:9]: It's for the program instruction. All types used there need to be Anchor/borsh-serializable. So using the `BigNumber` trait is not an option there

File: merkle-tree/indexed/Cargo.toml (hunk 1 above)
Comment (ananas-block) [reply merkle-tree/indexed/Cargo.toml:8]: ah I believe I was using 0.8.3

File: programs/address/src/lib.rs (hunk 3 above)
Comment (vadorovsky) [reply programs/address/src/lib.rs]: Borsh doesn't allow to use slices. We can use either `Vec` or an array with known size. Pick your poison. 💀 🍸

Comment (vadorovsky) [reply programs/address/src/lib.rs]: For now I just went with `Vec`.

File: programs/address/src/lib.rs (hunk 4 above)
Comment (vadorovsky) [reply programs/address/src/lib.rs]: Done.

File: pnpm-lock.yaml [hunk 5]
```diff
@@ -43,7 +45,7 @@ importers:
         version: 1.3.2
       '@rollup/plugin-typescript':
         specifier: ^11.1.5
-        version: 11.1.5(rollup@4.6.1)(typescript@5.3.2)
+        version: 11.1.5(rollup@4.6.1)(typescript@4.9.5)
```
Comment (sergeytimoshin) [inline pnpm-lock.yaml:48]: I think we need to revert this change. It introduced by running ./scripts/test.sh locally, and the reason is in the template: https://github.com/Lightprotocol/light-protocol/blob/main/psp-template/psp-template/package.json#L26

Comment (vadorovsky) [reply pnpm-lock.yaml:48]: OK. Any ideas how can I revert safely just that one change (downgrade of typescript) while keeping other relevant parts of the lock file?

Comment (ananas-block) [reply pnpm-lock.yaml:48]: you could try to upgrade the version in the psp-template

File: account-compression/Anchor.toml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+seeds = false
+
+[test.validator]
+url = "http://127.0.0.1:8899"
+
+[programs.localnet]
+psp_account_compression = "DmtCHY9V1vqkYfQ5xYESzvGoMGhePHLja9GQ994GKTTc"
+
+[programs.testnet]
+psp_account_compression = "DmtCHY9V1vqkYfQ5xYESzvGoMGhePHLja9GQ994GKTTc"
```
Comment (ananas-block) [inline account-compression/Anchor.toml:11]: Interesting that you put the account compression program into it's own anchor project. I would have put it into the other program folder, but I think it makes sense since this is a public good and we might want to put it into it's own repo eventually.

About naming, let's ditch the psp prefix since this will be not a light protocol utility but just the new compression program.

```suggestion
account_compression = "DmtCHY9V1vqkYfQ5xYESzvGoMGhePHLja9GQ994GKTTc"
```

File: account-compression/programs/account-compression/src/instructions/update_address_merkle_tree.rs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+pub fn process_update_address_merkle_tree<'info>(
+    ctx: Context<'_, '_, '_, 'info, UpdateMerkleTree<'info>>,
+    changelog_index: u16,
+    queue_index: u16,
+    address_index: u16,
+    address_next_index: u16,
+    address_next_value: [u8; 32],
+    low_address: RawIndexingElement<32>,
+    low_address_next_value: [u8; 32],
+    low_address_proof: [[u8; 32]; 40],
```
Comment (ananas-block) [inline account-compression/programs/account-compression/src/instructions/update_address_merkle_tree.rs:32]: I think this is not going to fit into one transaction.
2 * 4 + 32 + 66 + 32 + 40 * 32 = 1418

Comment (ananas-block) [reply account-compression/programs/account-compression/src/instructions/update_address_merkle_tree.rs:32]: We will need to add the zk proof (proving the Merkle proof) for the high element as well, which is an additional 128 bytes.

Comment (ananas-block) [reply account-compression/programs/account-compression/src/instructions/update_address_merkle_tree.rs:32]: I think the max we can do is height 23 if we want to do the update in one tx.
Otherwise we need to split it into two tx.
So let's split the update process into two tx (you can use the concept of the verifier state again).

**Review by ananas-block — COMMENTED** [review]
Thanks for the pr! Couple comments and considerations.

File: account-compression/programs/account-compression/src/instructions/update_address_merkle_tree.rs (hunk 7 above)
Comment (ananas-block) [reply account-compression/programs/account-compression/src/instructions/update_address_merkle_tree.rs:32]: Additional idea, you could already mock the zkp as input data with [u8;128] just to get the size right.

File: account-compression/Anchor.toml (hunk 6 above)
Comment (vadorovsky) [reply account-compression/Anchor.toml:11]: I can't put it in the same programs/ project, because that would cut off the possibility of using different Anchor CLIs.

If you move this program together with all the others, it's going to be picked up by `$TOP_DIR/programs/Anchor.toml`, therefore `light-anchor build` and `npx nx run @lightprotocol/program:build` are going to pick it too. And then IDL for all programs would be generated by one specific Anchor version.

Only havng separate Anchor.toml files guarantees what you wanted - possibility of using different Anchor versions, CLIs, IDLs.

**Comment by vadorovsky** [comment]
@ananas-block Are you fine with sticking to height 22 for this PR?

Changing it to 40 triggered a domino of issues I need to fix, to be precise:

* To have Merkle trees of height 40, we need to increase the size of zero bytes array to 40.
* Doing so blows up the stack memory.
* Switching zero bytes to `Vec<[u8; 32]>`... blew up the heap memory! 

```
[2024-02-11T15:30:01.124908074Z DEBUG solana_runtime::message_processor::stable_log] Program log: Instruction: InitializeAddressMerkleTree

vv my dumb prints

[2024-02-11T15:30:01.124919265Z DEBUG solana_runtime::message_processor::stable_log] Program log: aaaaa
[2024-02-11T15:30:01.124921719Z DEBUG solana_runtime::message_processor::stable_log] Program log: bbbbb
[2024-02-11T15:30:01.124923743Z DEBUG solana_runtime::message_processor::stable_log] Program log: ccccc
[2024-02-11T15:30:01.124925917Z DEBUG solana_runtime::message_processor::stable_log] Program log: qqq
[2024-02-11T15:30:01.124929895Z DEBUG solana_runtime::message_processor::stable_log] Program log: www

^^ end of my dumb prints, the last one is in the middle of `ConcurrentMerkleTree::init()`, after getting zero bytes and before processing changelog.

[2024-02-11T15:30:01.124976473Z DEBUG solana_runtime::message_processor::stable_log] Program log: Error: memory allocation failed, out of memory
```

I know it can be possibly solved with custom heap allocator, but I would need to make some changes to it, like:

* moving it to separate crate
* adjusting the immutable `&self` check (not doing it when the wrapped function is just a function, not a method)
* possibly something else which I didn't encounter yet

I'm worried that if I try to do it in this PR, I will blow it up and easily spend another 1-2 days just for that. I would prefer to sort out the memory issues in a separate PR. Are you fine with that?

File: programs/account-compression/src/instructions/initialize_address_merkle_tree.rs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+) -> Result<()> {
+    let mut address_merkle_tree = ctx.accounts.merkle_tree.load_init()?;
+    let address_merkle_tree =
+        address_merkle_tree_from_bytes_mut(&mut address_merkle_tree.merkle_tree);
+    address_merkle_tree
+        .init()
+        .map_err(|_| AccountCompressionErrorCode::AddressMerkleTreeInitialize)?;
+
+    msg!("changelog_index: {}", address_merkle_tree.changelog_index());
+    msg!("initialized tree succefully!!!111");
```
Comment (ananas-block) [inline programs/account-compression/src/instructions/initialize_address_merkle_tree.rs:25]: leftover log

File: programs/account-compression/src/instructions/insert_address.rs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    addresses: Vec<[u8; 32]>,
+) -> Result<()> {
+    let mut address_queue = ctx.accounts.queue.load_mut()?;
+    let address_queue = address_queue_from_bytes_mut(&mut address_queue.queue);
+
+    for address in addresses.iter() {
+        let address =
+            be_bytes_to_bigint(address).map_err(|_| AccountCompressionErrorCode::BytesToBigint)?;
+        address_queue
+            .append(address)
```
Comment (ananas-block) [inline programs/account-compression/src/instructions/insert_address.rs:26]: Have you checked how many comparisons we can do onchain?
This would determine whether we need to send a low nullifier index as indicator where to look for the low nullifier.

File: programs/account-compression/tests/tests.rs [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+                address_bundle.new_low_element.index,
+                address_bundle.new_element.value,
+            )
+            .unwrap();
+    }
+}
+
+/// Test insertion of addresses to the queue, dequeuing and Merkle tree update.
+#[tokio::test]
+async fn test_address_queue() {
```
Comment (ananas-block) [inline programs/account-compression/tests/tests.rs:296]: can you add a failing test as well?

File: programs/account-compression/src/instructions/update_address_merkle_tree.rs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+#[allow(clippy::too_many_arguments)]
+pub fn process_update_address_merkle_tree<'info>(
+    ctx: Context<'_, '_, '_, 'info, UpdateMerkleTree<'info>>,
+    // Index of the Merkle tree changelog.
+    changelog_index: u16,
+    // Index of the address to dequeue.
+    queue_index: u16,
+    // Index of the next address.
+    address_next_index: u16,
```
Comment (ananas-block) [inline programs/account-compression/src/instructions/update_address_merkle_tree.rs:30]: Next address index needs to be a bigger number since it has to hold at least the number of leaves.
I think u64 is best here.

For changelog_index and queue_index u16 is ok since it's constrained to the size of the changelog and queue.

File: programs/account-compression/src/instructions/insert_address.rs (hunk 9 above)
Comment (ananas-block) [reply programs/account-compression/src/instructions/insert_address.rs:26]: but for me we can do this in a different pr, so that have the anchor changes merged

File: utils/src/lib.rs [hunk 12]
```diff
@@ -29,6 +29,53 @@ pub fn change_endianness<const SIZE: usize>(bytes: &[u8; SIZE]) -> [u8; SIZE] {
     arr
 }
 
+/// Converts the given [`ark_ff::BigInt`](ark_ff::BigInt) into a big-endian
+/// byte array.
+pub fn bigint_to_be_bytes<const BYTES_SIZE: usize, const NUM_LIMBS: usize>(
```
Comment (ananas-block) [inline utils/src/lib.rs:34]: why don't you use the native ark ff BigInteger methods to_bytes_be and to_bytes_le?
https://docs.rs/ark-ff/latest/src/ark_ff/biginteger/mod.rs.html#448-461

**Review by ananas-block — CHANGES_REQUESTED** [review]
Awesome that it's green!! Found a leftover comment, no failing test and some other stuff. Adding a failing test is most important to me.
Once that's added happy to merge.

File: utils/src/lib.rs (hunk 12 above)
Comment (vadorovsky) [reply utils/src/lib.rs:34]: Because my function allocated an array with known size only once and writes limbs directly to it.

`to_bytes_be` allocates and writes to a vector, so then we would have to copy it again to an array (with `try_into()`). Which is unnecessary, because we know the size of the specific BigInt (through `NUM_LIBS` before).

I'm surprised that arkworks doesn't provide such method.

Comment (vadorovsky) [reply utils/src/lib.rs:34]: Actually, thanks for pointing out to this code. I looked at it again and I have one more optimization in mind. :smiling_imp:  Let me push it.

Comment (vadorovsky) [reply utils/src/lib.rs:34]: Done.

File: programs/account-compression/tests/tests.rs (hunk 10 above)
Comment (vadorovsky) [reply programs/account-compression/tests/tests.rs:321]: Done.

File: programs/account-compression/src/instructions/update_address_merkle_tree.rs (hunk 11 above)
Comment (vadorovsky) [reply programs/account-compression/src/instructions/update_address_merkle_tree.rs]: Changed to usize here, it seems to work.

Also, I made the index fields in `IndexingArray` and `IndexingElement` generic. You might get a heart attack when looking at how many traits I had to use, but welp, it works and it's probably still better than copying the whole struct definitions or using macros.

**Comment by vadorovsky** [comment]
@ananas-block PTAL

File: merkle-tree/indexed/tests/tests.rs [hunk 13]
```diff
@@ -61,23 +62,26 @@ fn program_update<H>(
...
 {
     // Remove the nullifier from the queue.
-    let mut nullifier = queue.dequeue_at(queue_index).unwrap().unwrap();
+    let nullifier = queue.dequeue_at(queue_index).unwrap().unwrap();
 
     // Update the nullifier with ranges adjusted to the Merkle tree state,
     // coming from relayer.
-    nullifier.index = nullifier_index;
-    nullifier.next_index = nullifier_next_index;
+    let nullifier: IndexingElement<usize, BigInteger256> = IndexingElement {
```
Comment (ananas-block) [inline merkle-tree/indexed/tests/tests.rs:80]: nice!

File: utils/src/lib.rs (hunk 12 above)
Comment (ananas-block) [reply utils/src/lib.rs:34]: nice! We should upstream at some point.

File: utils/src/lib.rs [hunk 14]
```diff
@@ -114,3 +162,88 @@ pub fn rustfmt(code: String) -> Result<Vec<u8>, anyhow::Error> {
...
+        BigInteger768, BigInteger832, UniformRand,
+    };
+    use rand::thread_rng;
+
+    use super::*;
+
+    const ITERATIONS: usize = 64;
+
+    #[test]
+    fn test_bigint_conversion() {
```
Comment (ananas-block) [inline utils/src/lib.rs:179]: What about edge cases, 0 to 0, MAX to MAX?
Have you ever run it with say 10k iterations?
(not saying that we should do 10k iterations in the standard test)

File: programs/account-compression/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+pub mod errors;
+pub mod instructions;
+use instructions::*;
+pub mod state;
+use state::*;
+
+declare_id!("5QPEJ5zDsVou9FQS3KCauKswM3VwBEBu4dpL9xTqkWwN");
+
+#[program]
+pub mod address {
```
Comment (ananas-block) [inline programs/account-compression/src/lib.rs:13]: should be account_compression right?

File: utils/src/lib.rs (hunk 14 above)
Comment (ananas-block) [reply utils/src/lib.rs:179]: Also pls add failing tests, in case I didn't miss them.

**Review by ananas-block — COMMENTED** [review]
Thanks for applying the changes!
Two nitpicks, name of the program module and edge case tests for your serialization.

File: programs/account-compression/src/lib.rs (hunk 15 above)
Comment (vadorovsky) [reply programs/account-compression/src/lib.rs]: Whoops, fixed

File: utils/src/lib.rs (hunk 14 above)
Comment (vadorovsky) [reply utils/src/lib.rs]: Added

**Review by ananas-block — APPROVED** [review]
